src/Raven.Server/Rachis/Follower.cs

```diff
@@ -411,7 +411,74 @@ private void NegotiateWithLeader(TransactionOperationContext context, LogLengthN
             // the reason we send this is to simplify the # of states in the protocol
 
             var snapshot = _connection.ReadInstallSnapshot(context);
-            _debugRecorder.Record("Snapshot received");
+            _debugRecorder.Record("Start receiving snapshot");
+
+            // reading the snapshot from network and committing it to the disk might take a long time. 
+            using (var cts = new CancellationTokenSource())
+            {
+                var readAndCommitSnapshotTask = Task.Run(() => ReadAndCommitSnapshot(context, snapshot, cts.Token), cts.Token);
+                try
+                {
+                    WaitForTaskCompletion(readAndCommitSnapshotTask, "ReadAndCommitSnapshot");
+                }
+                catch
+                {
+                    cts.Cancel();
+                    throw;
+                }
+            }
+
+            _debugRecorder.Record("Snapshot was received and committed");
+
+            // notify the state machine, we do this in an async manner, and start
+            // the operator in a separate thread to avoid timeouts while this is
+            // going on
+            var task = Task.Run(() => _engine.SnapshotInstalledAsync(snapshot.LastIncludedIndex));
```

Comment:
Do we need `cts` here as well?

Reply:
`SnapshotInstalledAsync` has two possible blocking calls.
1. opening a write tx
2. reload license

In both cases I'm not sure if cancellation is necessary. 
though we can cancel between the two calls.

Reply:
ok, not needed, then